src/main/scala/loamstream/model/execute/RxExecuter.scala

```diff
@@ -54,58 +55,111 @@ final case class RxExecuter(
-      //already ran 
-      (finishedJobs, notFinishedJobs) = jobs.partition(_.status.isTerminal)
-      if notFinishedJobs.nonEmpty
-      (jobsToMaybeRun, skippedJobs) = notFinishedJobs.map(_.job).partition(jobFilter.shouldRun)
-      (jobsToCancel, jobsToRun) = jobsToMaybeRun.partition(jobCanceler.shouldCancel)
-      _ = handleSkippedJobs(skippedJobs)
-      cancelledJobsMap = cancelJobs(jobsToCancel)
-      _ = record(jobOracle, cancelledJobsMap)
-      skippedResultMap = toSkippedResultMap(skippedJobs)
-      executionTupleOpt <- runJobs(jobsToRun, jobOracle)
-      _ = record(jobOracle, executionTupleOpt)
-      executionMap = cancelledJobsMap ++ executionTupleOpt
-      _ = logFinishedJobs(executionMap)
-    } yield {
-      executionMap ++ skippedResultMap
+    val chunkResults: Observable[Map[LJob, Execution]] = {
+      //Note onBackpressureDrop(), in case runEligibleJobs takes too long (or the polling window is too short)
+      val ticks = Observable.interval(windowLength, ioScheduler).onBackpressureDrop
+      
+      def runJobs() = runEligibleJobs(executionState, jobOracle)
+      
+      def isFinished = executionState.isFinished
+      
+      ticks.flatMap(_ => runJobs).takeUntil(_ => isFinished)
     }
-    //NB: We no longer stop on the first failure, but run each sub-tree of jobs as far as possible.
     
     val futureMergedResults = chunkResults.foldLeft(emptyExecutionMap)(_ ++ _).firstAsFuture
 
     Await.result(futureMergedResults, timeout)
   }
   
+  private def finish(executionState: ExecutionState)(results: Map[LJob, Execution]): Map[LJob, Execution] = {
+    def msg = {
+      val howMany: Int = scala.math.min(50, results.size) //scalastyle:ignore magic.number
+      
+      s"Finishing ${results.size} jobs; first $howMany ids: ${results.keys.take(howMany).map(_.id).mkString(",")}"
+    }
+    
+    TimeUtils.time(msg, debug(_)) {
+      executionState.finish(results)
+    }
+    
+    results
+  }
+  
+  private def runEligibleJobs(
+      executionState: ExecutionState, 
+      jobOracle: JobOracle): Observable[Map[LJob, Execution]] = {
+      
+    val jobsAndCells = executionState.updateJobs()
+    
+    val (numReadyToRun, numCannotRun) = (jobsAndCells.readyToRun.size, jobsAndCells.cannotRun.size)
+    import jobsAndCells.{ numRunning, numFinished }
+    val numRemaining = executionState.size - numReadyToRun - numCannotRun - numRunning - numFinished
+    
+    info(s"RxExecuter.runEligibleJobs(): Ready to run: $numReadyToRun; Cannot run: $numCannotRun; " +
+         s"Running: $numRunning; Finished: $numFinished; Other: $numRemaining.")
+    
+    val (finishedJobAndCells, notFinishedJobsAndCells) = {
+      jobsAndCells.readyToRun.partition { case (_, cell) => cell.isTerminal }
+    }
+    
+    val (finishedJobs, notFinishedJobs) = (finishedJobAndCells.keys, notFinishedJobsAndCells.keys)
+    
+    if(notFinishedJobs.nonEmpty) {
```

Comment:
I know this is a personal-preference thing, but when dealing with very long `if` statements that have an `else`, my gut feeling is that one of two things help w/ readability immensely:

1. Factor the body of the `if` into a function and call it. Possibly difficult in this situation given all the parameters it may have to take.

2. Do the `else` first and early-out with `return`:

```scala
if (notFinishedJobs.empty) {
  return Observable.just(Map.empty)
}

// now everything from here is fine and not part of a conditional...
```

Reply:
That's a fair point.  I'm not a fan of the early-return idiom in Scala, so I factored out a method.